src/blocks/packages.rs

```diff
@@ -0,0 +1,374 @@
+            }
+            PackageManager::Pacman => Box::new(Pacman::new().await?) as Box<dyn Backend>,
+            PackageManager::Aur => {
+                Box::new(Aur::new(config.aur_command.clone().unwrap_or_default()))
+                    as Box<dyn Backend>
+            }
+            PackageManager::Dnf => Box::new(Dnf::new()) as Box<dyn Backend>,
+        };
+
+        package_manager_vec.push(backend);
+    }
+
+    loop {
+        let mut package_manager_map: HashMap<&str, u32> =
+            [("apt", 0), ("pacman", 0), ("aur", 0), ("dnf", 0)].into();
+
+        let mut critical = false;
+        let mut warning = false;
+
+        // Iterate over the all package manager listed in Config
+        for package_manager in &package_manager_vec {
+            let mut updates = package_manager.get_updates_list().await?;
+            if let Some(regex) = ignore_updates_regex.clone() {
+                updates.retain(|u| !regex.is_match(u));
+            }
+
+            let updates_count = updates.len();
+
+            package_manager_map.insert(package_manager.name(), updates_count as u32);
+
+            warning |= warning_updates_regex
+                .as_ref()
+                .is_some_and(|regex| package_manager.has_matching_update(&updates, regex));
+            critical |= critical_updates_regex
+                .as_ref()
+                .is_some_and(|regex| package_manager.has_matching_update(&updates, regex));
+        }
+
+        let mut widget = Widget::new();
+
+        let total_count = package_manager_map.values().sum();
+        widget.set_format(match total_count {
+            0 => format_up_to_date.clone(),
+            1 => format_singular.clone(),
+            _ => format.clone(),
+        });
+        widget.set_values(map!(
+            "icon" => Value::icon("update"),
+            "apt" => Value::number(package_manager_map["apt"]),
+            "pacman" => Value::number(package_manager_map["pacman"]),
+            "aur" => Value::number(package_manager_map["aur"]),
+            "dnf" => Value::number(package_manager_map["dnf"]),
+            "total" => Value::number(total_count),
+        ));
+
+        widget.state = match total_count {
+            0 => State::Idle,
+            _ => {
+                if critical {
+                    State::Critical
+                } else if warning {
+                    State::Warning
+                } else {
+                    State::Info
+                }
+            }
+        };
+        api.set_widget(widget)?;
+
+        select! {
+            _ = sleep(config.interval.0) => (),
+            _ = api.wait_for_update_request() => (),
+        }
+    }
+}
+
+#[async_trait]
+pub trait Backend {
+    fn name(&self) -> &str;
+
+    async fn get_updates_list(&self) -> Result<Vec<String>>;
+
+    fn has_matching_update(&self, updates: &[String], regex: &Regex) -> bool {
+        updates.iter().any(|line| regex.is_match(line))
+    }
```

Comment:
Since none of the backends override this function, it doesn't use `self`, and I don't think any will in the future, let's move it out of the trait.

Reply:
Done